teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -130,6 +132,786 @@ Where they differ:
+[`emptyDir`](https://github.com/tektoncd/pipeline/blob/a7ad683af52e3745887e6f9ed58750f682b4f07d/docs/workspaces.md#emptydir)
+to be provided as a workspace at the Pipeline level even when that workspace is used by multiple Tasks, and when that
+happens, we take that as the cue to schedule those Tasks together.
+
+For example given this Pipeline:
+
+```yaml
+kind: Pipeline
+metadata:
+  name: build-test-deploy
+spec:
+ workspaces:
+  - name: source-code
+  - name: test-results
+ tasks:
+ - name: get-source
+   workspaces:
+   - name: source-code
+     workspace: source-code
+   taskRef:
+     name: git-clone
+   params:
+   - name: url
+      value: $(params.url)
+    - name: revision
+      value: $(params.revision)
+ - name: run-unit-tests
+   runAfter: get-source
+   taskRef:
+     name: just-unit-tests
+   workspaces:
+   - name: source-code
+     workspcae: source-code
+   - name: test-results
+     workspace: test-results
+ - name: upload-results
+   runAfter: run-unit-tests
+   taskRef:
+     name: gcs-upload
+   params:
+   - name: location
+     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+   workspaces:
+   - name: data
+     workspace: test-results
+```
+
+Running with this PipelineRun would cause `get-source` and `run-unit-tests` to be run in one pod, with `upload-results`
+in another:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: run
+spec:
+  pipelineRef:
+    name: build-test-deply
+  workspaces:
+  - name: source-code
+    emptyDir: {}
+  - name: test-results
+    persistentVolumeClaim:
+      claimName: mypvc
+```
+
+Running with this PipelineRun would cause all of the Tasks to be run in one pod:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: run
+spec:
+  pipelineRef:
+    name: build-test-deply
+  workspaces:
+  - name: source-code
+    emptyDir: {}
+  - name: test-results
+    emptyDir: {}
+```
+
+Running with this PipelineRun would cause all of the Tasks to be run in separate pods:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: run
+spec:
+  pipelineRef:
+    name: build-test-deply
+  workspaces:
+  - name: source-code
+    persistentVolumeClaim:
+      claimName: otherpvc
+  - name: test-results
+    persistentVolumeClaim:
+      claimName: mypvc
+```
+
+Pros:
+* Allows runtime decisions about scheduling without changing the Pod
+
+Cons:
+* If it's important for a Pipeline to be executed in a certain way, that information will have to be encoded somewhere
+  other than the Pipeline 
+
+#### PipelineRun: flag
+
+This is similar to the `emptyDir` based solution but instead of adding extra meaning to `emptyDir` we add a flag to the
+runtime workspace information or to the entire PipelineRun (maybe when this flag is set workspaces do not need to be
+provided.)
+
+#### Controller level
+
+This option is [TEP-0046](https://github.com/tektoncd/community/pull/318). In this option, the Tekton controller can
```

Comment:
Well, yes and no. This is suggested as an initial configuration in TEP-0046 because, as listed in Cons-section here, it addresses a problem that seem to require a more complex solution - Task parallelism - and therefore probably need to be iterated on multiple times.

Reply:
@jlpettersson can you elaborate a bit more? it sounds like you're saying a controller wide solution isnt necessarily the goal of TEP-0046? (or TEP-0046 is trying to say that a controller wide flag is a step toward a more complex solution?)

Reply:
The goal of TEP-0046 is to solve the Task parallelism-problem, that has been discussed back and forth since may 2020 and we later added the affinity assistant. The TEP links to a Design doc for the whole Task parallelism problem including alternative solutions (including affinity assistant and custom scheduler and pod-internal workspace). 

As has been noted, to solve that problem in the direction as TEP-0046 suggests, require a lot of changes, e.g. rethinking entrypoint and also propbably also the *run-CRDs (e.g. `PipelineRun`, `TaskRun` and perhaps `TaskGroupRun`) and the use of statuses and so on. This problem most likely needs to be solved in many iteration where we learn more after each, and the initial proposal is to do this behind a feature-flag or "Controller level flag" at least until we know more.